trieste/space.py

```diff
@@ -480,6 +502,175 @@ def __eq__(self, other: object) -> bool:
+            category_lens: Sequence[int] = categories  # type: ignore[assignment]
+            if any(x <= 0 for x in category_lens):
+                raise ValueError("Numbers of categories must be positive")
+            tags = [tuple(f"{i}" for i in range(n)) for n in category_lens]
+        else:
+            category_names: Sequence[Sequence[str]] = categories  # type: ignore[assignment]
+            if any(len(ts) == 0 for ts in category_names):
+                raise ValueError("Category name lists cannot be empty")
+            tags = [tuple(ts) for ts in category_names]
+
+        self._tags = tags
+
+        ranges = [tf.range(len(ts)) for ts in tags]
+        meshgrid = tf.meshgrid(*ranges, indexing="ij")
+        points = (
+            tf.reshape(tf.stack(meshgrid, axis=-1), [-1, len(tags)]) if tags else tf.zeros([0, 0])
+        )
+
+        super().__init__(points)
+
+    def __repr__(self) -> str:
+        """"""
+        return f"CategoricalSearchSpace({self._tags!r})"
+
+    @property
+    def has_bounds(self) -> bool:
+        return False
+
+    @property
+    def lower(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def upper(self) -> TensorType:
+        raise AttributeError("Categorical search spaces do not have numerical bounds")
+
+    @property
+    def tags(self) -> Sequence[Sequence[str]]:
+        """The tags of the categories."""
+        return self._tags
+
+    @property
+    def one_hot_encoder(self) -> EncoderFunction:
+        """A one-hot encoder for the numerical indices."""
+
+        def encoder(x: TensorType) -> TensorType:
+            if tf.rank(x) != 2:
+                raise ValueError(
+                    f"Invalid input for one-hot encoding: expected rank 2, got {tf.rank(x)}"
+                )
+            if x.shape[1] != len(self.tags):
+                raise ValueError(
+                    "Invalid input for one-hot encoding: "
+                    f"expected {len(self.tags)} tags, got {x.shape[1]}"
+                )
+            columns = tf.split(x, x.shape[1], axis=1)
+            encoders = [
+                tf.keras.layers.CategoryEncoding(num_tokens=len(ts), output_mode="one_hot")
+                for ts in self.tags
+            ]
+            return tf.concat(
+                [encoder(column) for encoder, column in zip(encoders, columns)], axis=1
+            )
+
+        return encoder
+
+    def to_tags(self, indices: TensorType) -> TensorType:
+        """
+        Convert a tensor of indices (such as one returned by :meth:`sample`) to one of
+        category tags.
+
+        :param indices: A tensor of integer indices.
+        :return: A tensor of string tags.
+        """
+
+        def extract_tags(row: TensorType) -> TensorType:
+            return tf.stack(
+                [tf.gather(tf.constant(self._tags[i]), row[i]) for i in range(len(row))]
+            )
```

Comment:
I _think_ you can probably use `tf.gather_nd` instead of `tf.gather` and a for loop. Not sure, but you might be able to also get rid of `tf.map_fn`.

Reply:
Not sure that's possible, as self._tags is a non-rectangular Sequence[Sequence[str], not a tensor.

Reply:
Okay :+1: .

There might be a way ([see example](https://stackoverflow.com/questions/67123407/tensorflow-how-to-use-a-ragged-tensor-as-an-index-into-a-normal-tensor)) to use ragged tensors with `batch_dims` arg to `tf.gather`, but that might not be more efficient than a `for` loop. Happy for you to ignore this :-)